### DISCUSSION TITLE
Any way to keep a TCP socket open longer than default?

### DISCUSSION THREAD

**Discussion by ma261065** [micropython/micropython#12225] [ESP32] [answered]
I have an audio player that streams audio from a server. 

If the user pauses the audio, the socket stops reading data.

I understand that "unused" TCP connections will be torn down after a period of time, by design.

However, if the user then resumes the audio after the TCP keepalive expires, the socket has already closed, and I receive a socket error.

Is there any way to keep a socket alive for longer?

I see that ESP-IDF has a SO_KEEPALIVE setting, but this isn't surfaced to MicroPython:

`sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 10000)`
`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'SO_KEEPALIVE'`

**Comment by jimmo** [comment, marked as answer]
It would be straightforward to add support for `SO_KEEPALIVE` as it's supported by LWIP (and as you note, the IDF uses LWIP, but we also use it on many other boards). See the implementation of `REUSEADDR`, it's much the same, just plumbing.

If you were interested in making a PR, this would need to be added to `extmod/modlwip.c`, `ports/esp32/modsocket.c`, and `ports/unix/modsocket.c` (and maybe Zephyr's ports/zephyr/modsocket.c too?).

(It's a bit frustrating that ESP32 has its own implementation of `socket` even though it's largely the same as the common LWIP one in modlwip -- it would be good to merge these one day)

**Reply by ma261065** [reply]
Thanks. I added the code to `ports/esp32/modsocket.c` to try it out for my use case, then called `s.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)`, which succeeded, but the socket still got closed after some (quite short) period of inactivity.

Do you know where the timeout is stored for closing down a socket? Is it configurable in menuconfig (I couldn't find it if it is)?
